### PR TITLE
motr_sync: [CORTX-33691] Define and implement rgw_motr synchronization

### PR DESCRIPTION
**Pull request by sumedhak27**
* Defines abstract synchronization primitive, with lock/unlock interface.
* Implements synchronization for Motr GC across multiple RGW instances.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/motr/sync/motr_sync.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+#include "rgw_sal_motr.h"
+
+// Define MOTR_EXCLUSIVE_RW_LOCK only when system needs exclusive read/write lock
+#define MOTR_EXCLUSIVE_RW_LOCK 1
+
+class MotrLockProvider;
+
+struct motr_locker_info_t
+{
+  std::string coockie;        // unique id of caller
```
Comment (jjxsg) [inline src/rgw/motr/sync/motr_sync.h:32]: What is `coockie`?

File: src/rgw/motr/sync/motr_sync_impl.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Steps:
+1. Follow step (1) in Approach 1.
+2. Call create_motr_Lock_instance(), passing it lock provider instance created
+in step (1) above.
+3. Call get_lock_instance() wherever lock is needed.
+
+Note: Presently, locking framework supports EXCLUSIVE lock; it means
+caller needs to set MotrLockType::EXCLUSIVE while calling  MotrLock::lock(),
+and do not pass locker_id.
+*/
```
Comment (jjxsg) [inline src/rgw/motr/sync/motr_sync_impl.h:66]:
```suggestion
/*

Usage:
    To use locks implemented in this file, the caller needs to instantiate
    class "MotrKVLockProvider" to create object. On such object, caller needs
    to call "initialize()" function, which in turn creates a global lock index,
    using the name specified by the caller. The caller can do this during the
    process/thread start-up/initialization; if "initialize()" returns < 0, the
    caller is expected to stop further initialization if locking is esential
    in the system. In multi-process environment, each process can call
    "MotrKVLockProvider::initialize" during it's start-up, providing the same
    name for the global lock index.

    After "MotrKVLockProvider" instance is created and initialized successfully,
    when the caller wants to acquire a resource, caller can first instantiate
    "MotrLock" class, passing it the instance of class "MotrKVLockProvider" in
    MotrLock::initialize() method. The caller can call MotrLock::lock() to lock
    the resource. Once the caller is done with the resource, caller can 
    call "MotrLock::unlock()". Any -ve value returned by MotrLock::lock()
    indicates resource can't be locked; a value 0 indicates success.

    Approach 1.
        Steps for caller (developer):
        1. During startup, create instance of class MotrKVLockProvider. Call
        initialize() on the instance, passing it a globally unique name for the
        global lock index creation. If initialize() fails (i.e. returns < 0),
        caller may proceed further or exit the application/system.
        2. Once step (1) is successful, create a global instance of class MotrLock,
        call MotrLock::initialize() and pass it KV lock provider instance
        created in step (1).
        3. Just before acquiring a resource, call MotrLock::lock() on the global
        instance of MotrLock.
        4. After work with the resource is done, release the lock by calling MotrLock::
        unlock() on the global instance of MotrLock.

    Approach 2.
        Steps:
        1. Follow step (1) in Approach 1.
        2. Call create_motr_Lock_instance(), passing it lock provider instance created
        in step (1) above.
        3. Call get_lock_instance() wherever lock is needed.
    
Note: Presently, locking framework supports EXCLUSIVE lock; it means
caller needs to set MotrLockType::EXCLUSIVE while calling  MotrLock::lock(),
and do not pass locker_id.

*/
```

**Review by jjxsg — APPROVED** [review]

File: src/rgw/motr/sync/motr_sync.h (hunk 1 above)
Comment (sumedhak27) [reply src/rgw/motr/sync/motr_sync.h:32]: coockie is identifier of an lock holder, we can call it `lock_holder` or `holder_id` as well. 
but to keep it generic @DPG17 has kept it as `cookie` I think.

Comment (jjxsg) [reply src/rgw/motr/sync/motr_sync.h:32]: Yes, it should be `cookie` and not `coockie`

Comment (sumedhak27) [reply src/rgw/motr/sync/motr_sync.h]: done

File: src/rgw/motr/sync/motr_sync_impl.h (hunk 2 above)
Comment (sumedhak27) [reply src/rgw/motr/sync/motr_sync_impl.h]: updated the Usage info. Please check.

Comment (jjxsg) [reply src/rgw/motr/sync/motr_sync_impl.h]: Looks good.

**Review by jjxsg — APPROVED** [review]

**Review by sachinpunadikar — APPROVED** [review]
LGTM

**Pull request merged**